keras/engine/training.py

```diff
@@ -2244,6 +2245,17 @@ def save_weights(self,
     if save_format == 'h5':
       with h5py.File(filepath, 'w') as f:
         hdf5_format.save_weights_to_hdf5_group(f, self.layers)
+        weights = self._trainable_weights + self._non_trainable_weights
+        weight_values = backend.batch_get_value(weights)
+        weight_names = [w.name.encode('utf8') for w in weights]
+        hdf5_format.save_attributes_to_hdf5_group(f, 'weight_names', weight_names)
+        for name, val in zip(weight_names, weight_values):
+          param_dset = f.create_dataset(name, val.shape, dtype=val.dtype)
+          if not val.shape:
+            # scalar
+            param_dset[()] = val
+          else:
+            param_dset[:] = val
```

Comment:
This code block is redundant with the logic of `save_weights_to_hdf5_group`. I would recommend extracting an abstract function that can be reused in both locations.